src/crystal/system/unix/io_uring.cr

```diff
@@ -225,20 +225,22 @@ class Crystal::System::IoUring
 
     # Populate fields and submit the index into the submission queue.
     yield sqe
-    sqe.value.user_data = user_data_pointer.address
+    sqe.value.user_data = pointerof(fiber).address
     sqe.value.opcode = opcode
 
     # If there is a timeout, we submit a LINK_TIMEOUT as well. It will make the
     # original event fail with ECANCELED if it doesn't complete in time.
-    unless timeval.null?
+    if timeout
       until (index_timeout = @submission_queue.consume_free_index) != UInt32::MAX
         process_completion_events(blocking: false)
       end
 
+      @timeval_entries[index_timeout] = make_timeval(timeout)
```

Comment:
Hmm. 

If I understand this logic correctly this depends on uring making a kernelside copy of the time structs, or else the server could consume the sqe and then userspace could emit get another sqe with a different timeout value, thus overwriting the old value. 

At least in older kernels that was not the case (for io_vec and msg_header at the very least). It is hopefully fixed now but early on userspace *had* to keep referred to structs alive and not overwrite them. I wonder if the timeout functions came before or after they started to be a bit more userfriendly.

Reply:
This is a very good catch and I was simply assuming the kernel was going to make a copy of this struct.

I went looking into the kernel source code and went back to 5.4 (when TIMEOUT was introduced) and 5.5 (when LINK_TIMEOUT was introduced) and luckily both of these do make a copy of the struct at submission time. That makes things easier!

Thanks for this insight.